packages/visual-editor/src/utils/getTreeDiff.ts

```diff
@@ -0,0 +1,195 @@
+import {
+  CompositionComponentNode,
+  CompositionTree,
+} from '@contentful/experience-builder-core/types';
+
+import { getItem } from './getItem';
+import { countNodes } from './treeHelpers';
+import { isEqual } from 'lodash-es';
+import { ROOT_ID } from '@/types/constants';
+import { TreeDiff } from '@/types/treeActions';
+
+interface MissingNodeActionParams {
+  index: number;
+  nodeAdded: boolean;
+  tree: CompositionTree;
+  child: CompositionComponentNode;
+  currentNode: CompositionComponentNode;
+  parentNodeId: string;
+}
+
+function missingNodeAction({
+  index,
+  nodeAdded,
+  child,
+  tree,
+  parentNodeId,
+  currentNode,
+}: MissingNodeActionParams): TreeDiff | null {
+  if (nodeAdded) {
+    return { type: 'add_node', indexToAdd: index, nodeToAdd: child, parentNodeId };
+  }
+
+  const item = getItem({ id: child.data.id }, tree);
+
+  if (item) {
+    const parentNode = getItem({ id: item.parentId! }, tree);
+
+    if (!parentNode) {
+      return null;
+    }
+    const sourceIndex = parentNode.children.findIndex((c) => c.data.id === child.data.id);
+
+    return { type: 'move_node', sourceIndex, destinationIndex: index, parentNodeId };
+  }
+
+  return {
+    type: 'replace_node',
+    originalId: currentNode.children[index].data.id,
+    node: child,
+  };
+}
+
+interface MatchingNodeActionParams {
+  index: number;
+  originalIndex: number;
+  nodeAdded: boolean;
+  nodeRemoved: boolean;
+  parentNodeId: string;
+}
+
+function matchingNodeAction({
+  index,
+  originalIndex,
+  nodeRemoved,
+  nodeAdded,
+  parentNodeId,
+}: MatchingNodeActionParams): TreeDiff | null {
+  if (index !== originalIndex && !nodeRemoved && !nodeAdded) {
+    return {
+      type: 'reorder_node',
+      sourceIndex: originalIndex,
+      destinationIndex: index,
+      parentNodeId,
+    };
+  }
+
+  return null;
+}
+
+interface CompareNodeParams {
+  currentNode?: CompositionComponentNode;
+  updatedNode?: CompositionComponentNode;
+  currentNodeCount: number;
+  updatedNodeCount: number;
+  originalTree: CompositionTree;
+  differences: Array<TreeDiff | null>;
+}
+
+function compareNodes({
```

Comment:
This is looking great and I can't wait to see those improvements in action!

If I'm not mistaken this is currently not covering updated entries/ assets that linked via `dataSource`. In [this recent PR](https://github.com/contentful/user_interface/pull/19090/files#diff-4e4cdedb6684dd71349bf556a416fe99ac358a33e84c3809f5ffa447339bd48aR50), I added the logic that would call `treeUpdated` to cause a rerendering as `updatedEntity` somehow didn't cause a rerendering.

Can we check here also if the linked entities in the entity store changed somehow? Otherwise, we would need to find out how to rerender the correct node on `updatedEntity` which is currently not working.

Just want to call this out as it looks like this change will break the feature that I shipped in that linked PR.

Reply:
@Chaoste Gotcha I'll test this out today and see how we could get that logic covered in this PR.

Just to clarify, the use case that currently doesn't work would be:
- I bind let's say a "header" text component to another entity
- I change the content of that entity and publish it
- I should see the changed content immediately in my experience

Reply:
@Chaoste just tested this scenario with both entries and assets:

- Linked text to a blog page entry, changed the content of the blog page, instantly saw the changes in the layout
- Linked an image to an existing asset, changed the file in the asset, instantly saw the changed file in the layout

No changes were required so this seems to function properly as is :)

The reason it works is a bit of a mystery to me. It does seem like the proper components are re-rendered but it's not because of a diff in the tree. I tried debugging a bit with [wdyr](https://github.com/welldone-software/why-did-you-render) but wasn't able to determine the reason. We may want to look into this more at some point